File: Java/dungeon-game.java
public class Solution {
    public int calculateMinimumHP(int[][] d) {
        int n = d.length;
        int m = d[0].length;
        int[][] mark = new int[n][m];
        mark[n-1][m-1] = Math.max(1, 1 - d[n-1][m-1]);
        for(int i=n-1; i>=0; i--) {
            for(int j=m-1; j>=0; j--) {
                int down = 0;
                int right = 0;
                if(i < n-1) {
                    down = Math.max(1, mark[i+1][j] - d[i][j]);
                }
                if(j < m-1) {
                    right = Math.max(1, mark[i][j+1] - d[i][j]);
                }
                if(down == 0 && right == 0) mark[i][j] = Math.max(1, 1 - d[i][j]);
                else {
                    if(down == 0) down = Integer.MAX_VALUE;
                    if(right == 0) right = Integer.MAX_VALUE;
                    mark[i][j] = Math.min(down, right);
                }
            }
        }
        return mark[0][0];
    }
}

File: Swift/dungeon-game.swift
/**
 * https://leetcode.com/problems/dungeon-game/
 * 
 * 
 */ 
// Date: Sun Jun 21 22:13:33 PDT 2020
class Solution {
    func calculateMinimumHP(_ dungeon: [[Int]]) -> Int {
        let n = dungeon.count
        guard let m = dungeon.first?.count else { return 0 }
        // minimum required health points to reach (x, y)
        // after costing certain points, it should be at least 1.
        var minHealth = Array(repeating: Array(repeating: 0, count: m), count: n)
        minHealth[n - 1][m - 1] = max(1, 1 - dungeon[n - 1][m - 1])
        
        for x in stride(from: n - 2, through: 0, by: -1) {
            minHealth[x][m - 1] = max(1, minHealth[x + 1][m - 1] - dungeon[x][m - 1])
        }
        for y in stride(from: m - 2, through: 0, by: -1) {
            minHealth[n - 1][y] = max(1, minHealth[n - 1][y + 1] - dungeon[n - 1][y])
        }
        
        for x in stride(from: n - 2, through: 0, by: -1) {
            for y in stride(from: m - 2, through: 0, by: -1) {
                let cost = min(minHealth[x + 1][y], minHealth[x][y + 1]) - dungeon[x][y]
                minHealth[x][y] = max(1, cost)
            }
        }
        // print("\(minHealth)")
        return minHealth[0][0]
    }
}